### DISCUSSION TITLE
Stepping in a multithreaded application

### DISCUSSION THREAD

**Discussion by pub-pub** [microsoft/debugpy#1761] [Enhancements]
i posted my initial issue [here](https://github.com/microsoft/vscode-python-debugger/issues/515) and it was successfully resolved.

now i'm using `"steppingResumesAllThreads": false` setting for debugging. but this solutions has a drawback, it pauses all other threads.

is it possible to add one more setting that will keep debugger Stepping inside my initial thread without pausing all other threads?

**Comment by rchiodo** [comment]
> is it possible to add one more setting that will keep debugger Stepping inside my initial thread without pausing all other threads?

I may be confused as to what you're asking but this is what stepping does by default? If you hit F10, all threads run, a breakpoint is put on the next line in the current thread, and then the breakpoint hits. All threads pause again.

**Reply by pub-pub** [reply]
> I may be confused as to what you're asking but this is what stepping does by default? If you hit F10, all threads run, a breakpoint is put on the next line in the current thread, and then the breakpoint hits. All threads pause again.

with `"steppingResumesAllThreads": false` all other threads are paused when i'm stepping through my 'debugged' thread, i see it because i have `time.sleep(10)` in 'debugged' thread and during that time all other threads just wait and do nothing. and for me it's ok.

i was asking about something a bit different. is it possible to make so that all other threads do not pause when i'm stepping through my 'debugged' thread?

**Reply by rchiodo** [reply]
That's what normal stepping does though. Set `"steppingResumesAllThreads": true` and it will resume other threads as you step.

**Reply by rchiodo** [reply]
It just stops them all when you actually end up at the next step location.

**Reply by rchiodo** [reply]
If you're asking whether or not you could have other threads running while stopped in one thread, sorry but that's not possible. Other threads can be writing to shared memory, taking the GIL etc, while your stopped thread is trying to eval the value of variables in the stack.

**Reply by rchiodo** [reply]
Your original issue made it sound like normal stepping isn't working. Step over should land in the same thread. There would be a bug in the stepping code if it ended up in a different thread.

**Reply by pub-pub** [reply]
> That's what normal stepping does though. Set `"steppingResumesAllThreads": true` and it will resume other threads as you step.

this was my initial issue that i reported [here](https://github.com/microsoft/vscode-python-debugger/issues/515#issue-2726887202)
the problem is that with `"steppingResumesAllThreads": true` stepping does not stick to the initial 'debugged' thread but randomly jumps throughout all threads.
a more detailed explanation can be found [here](https://developercommunity.visualstudio.com/t/make-the-debugger-stick-to-the-current-thread-inst/351390), although it is not vs code, but the problem is the same.

**Reply by pub-pub** [reply]
> If you're asking whether or not you could have other threads running while stopped in one thread, sorry but that's not possible. Other threads can be writing to shared memory, taking the GIL etc, while your stopped thread is trying to eval the value of variables in the stack.

i was suspecting something like that, but thanks anyway

**Reply by rchiodo** [reply]
Are your threads all running the same code? That would explain why it would jump around. Unlike Visual Studio's C++ debugger, debugpy doesn't have a way to tie a breakpoint to just a specific thread.

**Reply by pub-pub** [reply]
> Are your threads all running the same code? That would explain why it would jump around. Unlike Visual Studio's C++ debugger, debugpy doesn't have a way to tie a breakpoint to just a specific thread.

no, my 'debugged' thread is different, it is some kind of 'watcher' thread, all other threads are workers and run the same code

**Reply by rchiodo** [reply]
Hmm, then that sounds like a stepping issue. It shouldn't step into random code. It's supposed to step to the next line that's going to execute.

**Reply by pub-pub** [reply]
> Hmm, then that sounds like a stepping issue. It shouldn't step into random code. It's supposed to step to the next line that's going to execute.

when i set `"steppingResumesAllThreads": false` it stays inside my 'debugged' thread, but when it's `true` it jumps outside

**Reply by rchiodo** [reply]
Well yeah because the other threads are paused, there's no new event for another line executing from any thread but the one you have paused.

When the next event comes in it should be checking if it's the step over/out location. And if it is, firing a breakpoint event. If the other thread isn't running the same code, then that location shouldn't be a possible step over/out location.

Do you have a sample app that reproduces the issue?

**Reply by pub-pub** [reply]
> Do you have a sample app that reproduces the issue?

no, i'm working on the 'watcher' thread only,
but i will try to write smth short and see if i can reproduce the issue

**Comment by pub-pub** [comment]
sorry i don't know how to post source code here, so i attached it as txt file

on the screenshot you can see that my breakpoint is on line 12
now run the code and when it stops, press F10 five times and it will jump to line 7

when i was writing this sample i've noticed one strange thing.
when you uncomment line 6, the debugger will run fine and will not jump outside 'watcher' thread.


![image](https://github.com/user-attachments/assets/0ef9dce1-d177-4088-981f-cf8042b5b495)

[Untitled-1.txt](https://github.com/user-attachments/files/18087475/Untitled-1.txt)

**Reply by rchiodo** [reply]
Thanks, I can reproduce the problem. Seems incorrect to me. I'm opening an issue.

**Reply by rchiodo** [reply]
https://github.com/microsoft/debugpy/issues/1765
